### ISSUE TITLE
incorrect results for lodash

### ISSUE THREAD

**Issue by jdalton** [kangax/jscritic#1]
I tried jscritic on [lodash](https://github.com/lodash/lodash/blob/master/lodash.js) and it incorrectly stated that it is extending native `Object`, fails to detect lodash's use of ES6 Set, incorrectly asserts that lodash is creating a global `global`, and incorrectly lists lots of used variables as unused.

**Comment by jdalton** [comment]
I narrowed down some of the issues:
- It seems to think I'm augmenting `Object` because I have a local:
  
  ``` js
  function Object() { ...}
  ```
  
  in the code. I do this on purpose so in dev consoles obects created by it show up as `Object`.
- It sees vars declared in if-statements as being used only once which is incorrect as they are hoisted and used in other places:
  
  ``` js
  if (typeof objLength == 'number' && objLength > 0) {
  var allowIndexes = isArray(object) || (support.unindexedChars && isString(object)),
        maxIndex = objLength - 1;
  }
  ```
- The `global` global var use is
  `var freeGlobal = freeExports && freeModule && typeof global == 'object' && global;`<br>
  which is valid because I type check before accessing.

**Comment by kentcdodds** [comment]
@jdalton, you rock dude. Love to see you making sure lodash is amazing everywhere.

**Comment by kangax** [comment]
:) @jdalton is the man!

**Comment by kangax** [comment]
Filed https://github.com/jshint/jshint/issues/1600 for 1st issue. Will look at others soon.

**Comment by kangax** [comment]
Fixed 2nd issue with https://github.com/kangax/jscritic/commit/15957a2e2a99607e328d945e9e898a845f76a22e

**Comment by kangax** [comment]
As for 3rd issue, it's the way JSHint detects so-called "implied" globals:

``` js
if (typeof foo === 'undefined' && foo) { .. }
```

^ here it catches `foo`, for example

**Comment by jdalton** [comment]
> As for 3rd issue, it's the way JSHint detects so-called "implied" globals

Maybe it's a messaging thing as with JSCritic I thought it was showing me undeclared var assignments (bigger worry) but its showing potential access of global vars (less of a worry).

**Comment by kangax** [comment]
Yeah. Perhaps I should remove implieds from "global vars" entirely.

**Comment by jdalton** [comment]
The other items left are the failure to detect ES6 Set. I see jshint has something about it [here](https://github.com/jshint/jshint/blob/10ea5f74f8fe156dcfa42cf9a0140b4487af1e6a/src/vars.js#L42-L49). 

It's also failing to detect inconsequential unused vars:

``` js
(function() {
  [1,2,3].forEach(function(value, index) {
    return index % 2;
  });
}());
// => detects `value` as unused, but only if wrapped in the IIFE
```

**Comment by kangax** [comment]
Filed this as https://github.com/jshint/jshint/issues/1605
